### PR TITLE
Move ws image buffer to ws object

### PR DESCRIPTION
**Pull request by TheNeikos**

### CODE REVIEW

**Comment by neithernut** [comment]
Targets #135.

**Comment by neithernut** [comment]
You have an ACK from me, but I'll wait for @matthiasbeyer.

File: src/compositor/background_surface.c [hunk 1]
```diff
@@ -5,13 +5,83 @@
...
+
+ws_buffer_type_id WS_OBJECT_TYPE_ID_IMAGE_BUFFER = {
+    .type = {
+        .supertype  = (ws_object_type_id*)&WS_OBJECT_TYPE_ID_BUFFER,
+        .typestr    = "ws_image_buffer",
+
+        .hash_callback = NULL,
+
+        .init_callback = NULL,
+        .deinit_callback = free_buffer,
```
Comment (matthiasbeyer) [inline src/compositor/background_surface.c]: I'd prefer the callback function named as the callback, so `free_buffer` should be `deinit_buffer` or `deinit_callback`.

File: src/compositor/background_surface.c [hunk 2]
```diff
@@ -41,3 +111,79 @@ ws_background_service_load_image(
     return buff;
 }
 
+struct ws_image_buffer*
+ws_image_buffer_new(
+    void
```
Comment (matthiasbeyer) [inline src/compositor/background_surface.c]: notation can be in one line if we have no arguments.

File: src/compositor/background_surface.c [hunk 3]
```diff
@@ -41,3 +111,79 @@ ws_background_service_load_image(
...
+    int cmp = strcmp(buff1->path, buff2->path);
+    return signum(cmp);
+}
+
+static void*
+get_data(
+    struct ws_buffer* self
+) {
+    struct ws_image_buffer* buff = (struct ws_image_buffer*)self;
+    return buff->buffer;
```
Comment (matthiasbeyer) [inline src/compositor/background_surface.c:155]: No check if `buff` is `NULL`?

File: src/compositor/background_surface.c [hunk 4]
```diff
@@ -41,3 +111,79 @@ ws_background_service_load_image(
...
+}
+
+static int32_t
+get_height(
+    struct ws_buffer* self
+) {
+    struct ws_image_buffer* buff = (struct ws_image_buffer*)self;
+    return  buff->height;
```
Comment (matthiasbeyer) [inline src/compositor/background_surface.c:163]: No check if `buff` is `NULL`?

File: src/compositor/background_surface.c [hunk 5]
```diff
@@ -41,3 +111,79 @@ ws_background_service_load_image(
...
+}
+
+static int32_t
+get_width(
+    struct ws_buffer* self
+) {
+    struct ws_image_buffer* buff = (struct ws_image_buffer*)self;
+    return buff->width;
```
Comment (matthiasbeyer) [inline src/compositor/background_surface.c:171]: No check if `buff` is `NULL`?

File: src/compositor/background_surface.c [hunk 6]
```diff
@@ -41,3 +111,79 @@ ws_background_service_load_image(
...
+}
+
+static int32_t
+get_stride(
+    struct ws_buffer* self
+) {
+    struct ws_image_buffer* buff = (struct ws_image_buffer*)self;
+    return buff->stride;
```
Comment (matthiasbeyer) [inline src/compositor/background_surface.c:179]: No check if `buff` is `NULL`?

File: src/compositor/background_surface.c (hunk 3 above)
Comment (TheNeikos) [inline src/compositor/background_surface.c:155]: Well, if self is null what am I supposed to return? You can't get `data` from a null object.

File: src/util/arithmetical.h [hunk 7]
```diff
@@ -34,7 +34,7 @@
 #ifndef __WS_UTIL_ARITHMETICAL_H__
 #define __WS_UTIL_ARITHMETICAL_H__
 
-
+#define signum(x) (((x) > 0) - ((x) < 0))
```
Comment (matthiasbeyer) [inline src/util/arithmetical.h:42]: Documentation missing.

Comment (neithernut) [inline src/util/arithmetical.h:42]: Obvious math is obvious.

File: src/compositor/background_surface.c (hunk 3 above)
Comment (matthiasbeyer) [inline src/compositor/background_surface.c:155]: `NULL`.

File: src/util/arithmetical.h (hunk 7 above)
Comment (matthiasbeyer) [inline src/util/arithmetical.h:42]: But at least a documentation like "macro for doing xyz" _should_ be there, so it appears in the documentation.

**Comment by neithernut** [comment]
Please rebase one #144 is merged and set the new flag `WS_OBJECT_HEAPALLOCED`.

File: src/util/arithmetical.h [hunk 8]
```diff
@@ -34,7 +34,11 @@
 #ifndef __WS_UTIL_ARITHMETICAL_H__
 #define __WS_UTIL_ARITHMETICAL_H__
 
-
+/* Returns the sign of the number put in
```
Comment (matthiasbeyer) [inline src/util/arithmetical.h]: This is not a doxygen comment. Style:

``` C
/**
 * Header
 *
 * Detailed
 */
```

File: src/util/arithmetical.h [hunk 9]
```diff
@@ -34,7 +34,12 @@
 #ifndef __WS_UTIL_ARITHMETICAL_H__
 #define __WS_UTIL_ARITHMETICAL_H__
 
-
+/*
```
Comment (matthiasbeyer) [inline src/util/arithmetical.h]: Comments for doxygen should start with `/**`, not `/*`.

**Comment by matthiasbeyer** [comment]
ACK.

**Comment by neithernut** [comment]
ACK.

**Pull request merged**